src/main/scala/com/cognite/sdk/scala/common/OAuth2.scala

```diff
@@ -65,106 +153,52 @@ object OAuth2 {
-                  F.raiseError(
-                    new SdkException(
-                      s"Failed to parse response from IdP: ${error.getMessage}"
-                    )
-                  )
-              }
+          now <- clock.monotonic
+          res <- maybeCacheToken match {
+            case Some(originalToken)
+                if now.toSeconds < (originalToken.expiresAt - refreshSecondsBeforeTTL) =>
+              F.delay(originalToken)
+            case _ =>
+              credentials.getAuth()
           }
-          acquiredAt <- clock.monotonic
-          expiresAt = acquiredAt.toSeconds + payload.expires_in - refreshSecondsBeforeTTL
-        } yield TokenState(payload.access_token, expiresAt, credentials.cdfProjectName)
-      }
-
+        } yield res
       ConcurrentCachedObject(authenticate).map(new ClientCredentialsProvider[F](_))
     }
   }
   // scalastyle:on method.length
 
   object SessionProvider {
-
-    /** Only use for SessionProvider to interact with Cognite internal SessionAPI */
-    private def getKubernetesJwt[F[_]](implicit F: Async[F]): F[String] = {
-      val serviceAccountTokenPath = Path("/var/run/secrets/tokens/cdf_token")
-      Files[F]
-        .readAll(serviceAccountTokenPath)
-        .through(fs2.text.utf8.decode)
-        .compile
-        .string
-        .adaptError { case err =>
-          new SdkException(s"Failed to get service token because ${err.getMessage}")
-        }
-    }
-
     def apply[F[_]](
         session: Session,
         refreshSecondsBeforeTTL: Long = 30,
-        getToken: Option[F[String]] = None
+        getToken: Option[F[String]] = None,
+        maybeCacheToken: Option[TokenState] = None
     )(
         implicit F: Async[F],
         clock: Clock[F],
         sttpBackend: SttpBackend[F, Any]
     ): F[SessionProvider[F]] = {
-      import sttp.client3.circe._
-      val authenticate: F[TokenState] = {
-        val uri = uri"${session.baseUrl}/api/v1/projects/${session.cdfProjectName}/sessions/token"
+      val authenticate: F[TokenState] =
         for {
-          kubernetesServiceToken <- getToken.getOrElse(getKubernetesJwt)
-          payload <- basicRequest
-            .header("Accept", "application/json")
-            .header("Authorization", s"Bearer $kubernetesServiceToken")
-            .post(uri)
-            .body(RefreshSessionRequest(session.sessionId, session.sessionKey))
-            .response(
-              parseResponse[SessionTokenResponse, SessionTokenResponse](uri, value => value)
-            )
-            .send(sttpBackend)
-            .map(_.body)
-          acquiredAt <- clock.monotonic
-          expiresAt = acquiredAt.toSeconds + payload.expiresIn - refreshSecondsBeforeTTL
-        } yield TokenState(payload.accessToken, expiresAt, session.cdfProjectName)
-      }
-
-      ConcurrentCachedObject(authenticate).map(new SessionProvider[F](_))
+          now <- clock.monotonic
+          res <- maybeCacheToken match {
```

Comment:
Same as above, use something more descriptive than "res" please.

Reply:
changed to `newToken` :)